src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +57,33 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+    std::filesystem::path path(key);
+
+    while (std::filesystem::is_empty(path) && path != _cache_dir) {
```

Comment:
unfortunately `is_empty` is probably a blocking system call (e.g. in cases where the kernel doesn't have inode cached). there is a way to do this with seastar via `seastar::file::list_directory`. I would recommend adding a small utility, perhaps along side the directory walker that we have in tree (`src/v/utils/directory_walker.h`).

Reply:
could you make use of `include/seastar/util/file.hh:future<> recursive_remove_directory(std::filesystem::path path) noexcept;` which might simplify things.

Reply:
actually `ss::remove_file` deletes only empty directory and throws an exception if the directory is not empty. Deleted the `std::filesystem::is_empty` check